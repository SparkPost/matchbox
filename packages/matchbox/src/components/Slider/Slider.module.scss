@import '../../styles/config';

.Slider {
  position: relative;
  margin-bottom: spacing();
  padding: spacing() 0;
  cursor: pointer;

  &.Disabled {
    cursor: not-allowed;
  }
}

.Rail {
  position: absolute;
  left: 0;
  right: 0;

  height: 4px;
  background: color(gray, 8);
  border-radius: 3px;
  transition: 0.3s background;

  .Slider:hover & {
    background: color(gray, 7);
  }
}

.Track {
  position: absolute;
  left: 0;
  height: 4px;
  background: color(orange);
  border-radius: 3px;

  .Slider.Disabled  & {
    background: color(gray, 4);
  }
}

.Handle {
  position: absolute;
  width: 15px;
  height: 15px;
  transform: translate(-7px, -5px);

  border-radius: 50%;
  border: 3px solid color(orange);
  background: color(gray, 10);
  transition: 0.1s box-shadow;

  outline: none;

  .Slider:not(.Disabled) & {
    &:focus, &:active, &:hover {
      outline: none;
      box-shadow: 0 0 0 5px rgba(color(orange), 0.3);
    }
  }

  .Slider.Disabled  & {
    border: 2px solid color(gray, 4);
  }
}

.Tick {
  position: absolute;
  height: 10px;
  width: 10px;
  border-radius: 50%;
  transform: translate(-3px, -3px);

  background: color(gray, 8);
  transition: 0.3s background;

  user-select: none;

  .Slider.Disabled &.Included {
    background: color(gray, 4);
  }

  .Slider:not(.Disabled) &.Included {
    background: color(orange);
  }

  .Slider:hover:not(.Disabled) &:not(.Included) {
    background: color(gray, 7);
  }
}

.TickLabel {
  position: absolute;
  top: 15px;
  left: 50%;
  transform: translate(-50%, 0);

  color: color(gray, 4);
  font-size: font-size(200);
}
