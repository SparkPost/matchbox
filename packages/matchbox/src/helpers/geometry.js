import React from 'react';
import { findDOMNode } from 'react-dom';
import { clamp } from './math';
import { debounce } from './event';

export function getWindowRect() {
  return {
    top: window.scrollY,
    left: window.scrollX,
    height: window.innerHeight,
    width: window.innerWidth
  };
}

export function getRectFor(node) {
  const rect = findDOMNode(node);

  if (!rect) {
    return {};
  }

  return rect.getBoundingClientRect();
}

/**
 * Provides a hook that returns window dimensions
 * @param  {number} [wait=100] Timer for debounced dimension calculation
 * @return {Shape}             Same results as `getWindowRect``
 */
export function useWindowSize(wait = 100) {
  const [size, setSize] = React.useState(getWindowRect());

  const handleResize = debounce(() => {
    setSize(getWindowRect());
  }, wait);

  React.useEffect(() => {
    window.addEventListener('resize', handleResize);
    return () => {
      window.removeEventListener('resize', handleResize);
    };
  }, []);

  return size;
}

/**
 * Gets preferred direction for the provided react component
 * @param  {React Node} node - a react component
 * @return {Shape}
 * Returns directional booleans for where the component should render. Eg:
 * top: true if component is in the bottom half of the screen
 * right: true if component in the left half of the screen
 */
export function getPreferredDirectionFor(node) {
  const windowRect = getWindowRect();
  const elementRect = getRectFor(node);

  const rightOffset = windowRect.width - elementRect.right;
  const bottomOffset = windowRect.height - elementRect.bottom;

  return {
    top: bottomOffset < elementRect.top,
    left: rightOffset < elementRect.left,
    right: rightOffset >= elementRect.left,
    bottom: bottomOffset >= elementRect.top
  };
}


/**
 * Gets coordinates and dimensions in pixels for the provided react component
 * @param  {React Node} node
 * @return {Shape}
 */
export function getPositionFor(node, { fixed = false } = {}) {
  const windowRect = getWindowRect();
  const elementRect = getRectFor(node);

  return {
    top: elementRect.top + (fixed ? 0 : windowRect.top),
    left: elementRect.left + (fixed ? 0 : windowRect.left),
    width: elementRect.width,
    height: elementRect.height
  };
}

/**
 * Linearly interpolates and clamps between two values
 * @param  {number} min
 * @param  {number} max
 * @param  {number} n
 * @return {number}
 * @example
 *   lerp(10, 20, 0.5)
 *   > 15
 */
export function lerp(min, max, n) {
  const value = (max - min) * n + min;
  return clamp(value, min, max);
}
