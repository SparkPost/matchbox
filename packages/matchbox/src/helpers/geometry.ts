import React from 'react';
import { findDOMNode } from 'react-dom';
import { clamp } from './math';
import { debounce } from './event';
import { getWindow } from './window';

export type WindowRect = {
  top: number;
  left: number;
  height: number;
  width: number;
};

export function getWindowRect(): WindowRect {
  const environment = getWindow();
  return {
    top: environment.scrollY,
    left: environment.scrollX,
    height: environment.innerHeight,
    width: environment.innerWidth,
  };
}

export function getRectFor(node?: React.ReactInstance): ClientRect {
  // Need to update this soon since findDomNode is deprecated
  const rect = findDOMNode(node) as Element; //eslint-disable-line

  if (!rect) {
    return {} as ClientRect;
  }

  return rect.getBoundingClientRect();
}

/**
 * Provides a hook that returns window dimensions
 * @param  {number} [wait=100] Timer for debounced dimension calculation
 * @return {Shape}             Same results as `getWindowRect``
 */
export function useWindowSize(wait = 100): WindowRect {
  const environment = getWindow();
  const [size, setSize] = React.useState(getWindowRect());

  const handleResize = debounce(() => {
    setSize(getWindowRect());
  }, wait);

  React.useEffect(() => {
    environment.addEventListener('resize', handleResize);
    return () => {
      environment.removeEventListener('resize', handleResize);
    };
  }, []);

  return size;
}

/**
 * Gets preferred direction for the provided react component
 * @param  {React Node} node - a react component
 * @return {Shape}
 * Returns directional booleans for where the component should render. Eg:
 * top: true if component is in the bottom half of the screen
 * right: true if component in the left half of the screen
 */
export function getPreferredDirectionFor(node?: React.ReactInstance): {
  top: boolean;
  left: boolean;
  right: boolean;
  bottom: boolean;
} {
  const windowRect = getWindowRect();
  const elementRect = getRectFor(node);

  const rightOffset = windowRect.width - elementRect.right;
  const bottomOffset = windowRect.height - elementRect.bottom;

  return {
    top: bottomOffset < elementRect.top,
    left: rightOffset < elementRect.left,
    right: rightOffset >= elementRect.left,
    bottom: bottomOffset >= elementRect.top,
  };
}

/**
 * Gets coordinates and dimensions in pixels for the provided react component
 * @param  {React Node} node
 * @return {Shape}
 */
export function getPositionFor(node?: React.ReactInstance, { fixed = false } = {}): WindowRect {
  const windowRect = getWindowRect();
  const elementRect = getRectFor(node);

  return {
    top: elementRect.top + (fixed ? 0 : windowRect.top),
    left: elementRect.left + (fixed ? 0 : windowRect.left),
    width: elementRect.width,
    height: elementRect.height,
  };
}

/**
 * Linearly interpolates and clamps between two values
 * @param  {number} min
 * @param  {number} max
 * @param  {number} n
 * @return {number}
 * @example
 *   lerp(10, 20, 0.5)
 *   > 15
 */
export function lerp(min: number, max: number, n: number): number {
  const value = (max - min) * n + min;
  return clamp(value, min, max);
}

/**
 * Rounds a number to the nearest baseline ceiling
 */
export function roundToBaseline(n: number, base = 4): number {
  return Math.ceil(n / base) * base;
}
