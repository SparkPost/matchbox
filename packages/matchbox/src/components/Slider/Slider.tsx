import React from 'react';
import styled from 'styled-components';
import { margin, MarginProps, compose } from 'styled-system';
import { getRectFor, lerp } from '../../helpers/geometry';
import { noop, isNotTouchEvent } from '../../helpers/event';
import { onKey, onKeys } from '../../helpers/keyEvents';
import { roundToPlaces, clamp } from '../../helpers/math';
import { getWindow } from '../../helpers/window';
import { useResizeObserver } from '../../hooks';
import {
  slider,
  rail,
  railHover,
  track,
  tick,
  tickLabel,
  tickLabelPosition,
  handle,
  handleShadow,
} from './styles';
import { pick } from '@styled-system/props';

const system = compose(margin);

type SliderStyledProps = {
  $included?: boolean;
  $disabled?: boolean;
};

export const StyledSlider = styled('div')`
  ${slider}
  ${system}
`;

const StyledRail = styled('div')<SliderStyledProps>`
  ${rail}
  ${StyledSlider}:hover & {
    ${railHover}
  }
`;

const StyledTrack = styled('div')<SliderStyledProps>`
  ${track}
`;

const StyledTick = styled('div')<SliderStyledProps>`
  ${tick}
`;

const StyledTickLabel = styled('div')`
  ${tickLabel}
  ${tickLabelPosition}
`;

const StyledHandle = styled('div')<SliderStyledProps>`
  ${handle}
`;

const StyledHandleShadow = styled('div')<SliderStyledProps>`
  ${handleShadow}
`;

export type SliderProps = {
  id?: string;
  /**
   * The slider's initial value on first render
   */
  defaultValue?: number;
  /**
   * Disables focus, key down, mouse and touch events
   */
  disabled?: boolean;
  /**
   * The slider's lower bounds
   */
  min?: number;
  /**
   * The slider's upper bounds
   */
  max?: number;
  onBlur?: React.FocusEventHandler<HTMLDivElement>;
  onChange?: (value: number) => void;
  onFocus?: React.FocusEventHandler<HTMLDivElement>;
  /**
   * The number of decimal places to round values to
   */
  precision?: number;

  /**
   * Generates tick marks
   */
  ticks?: { [k: number]: React.ReactNode };
  /**
   * A value to programatically control the slider
   */
  value?: number;
  /**
   * Describes a side-effect relationship with another DOM element
   */
  'aria-controls'?: string;
  'aria-labelledby'?: string;
  /**
   * Identifier passed to the handle for testing or tracking purposes
   */
  'data-id'?: string;
  /**
   * System props for margin
   */
} & MarginProps;

function Slider(props: SliderProps): JSX.Element {
  const {
    defaultValue,
    disabled,
    id,
    max,
    min,
    onBlur,
    onFocus,
    onChange,
    precision,
    ticks,
    value,
    ...rest
  } = props;

  const systemProps = pick(rest);
  const environment = getWindow();

  const [resizeRef, { contentRect }] = useResizeObserver();
  const [sliderValue, setSliderValue] = React.useState(
    value || defaultValue != null ? defaultValue : min,
  );
  const [sliderLocation, setSliderLocation] = React.useState(0);
  const [rect, setRect] = React.useState({ width: 0, left: 0 });
  const [moving, setMoving] = React.useState('');
  const sliderRef = React.useRef();

  React.useEffect(() => {
    setRect(getRectFor(sliderRef.current));
  }, [contentRect, sliderRef.current]);

  // Calculates step increments based on precision
  const interval = React.useMemo(() => {
    let interval = 1;
    if (precision > 0) {
      const zeros = '0'.repeat(precision - 1);
      interval = parseFloat(`0.${zeros}1`);
    }
    return interval;
  }, [precision]);

  // Sets internal value when value is controlled externally
  React.useEffect(() => {
    if (!isNaN(value) && isFinite(value)) {
      setValue(value);
    }
  }, [value]);

  // Updates slider location when value changes
  React.useEffect(() => {
    if (rect.width) {
      const absoluteProportion = (sliderValue - min) / Math.abs(min - max);
      setSliderLocation(lerp(0, rect.width, absoluteProportion));
      if (onChange) {
        onChange(sliderValue);
      }
    }
  }, [sliderValue, rect.width]);

  // Updates tick locations when ticks or component size change
  const tickLocations = React.useMemo(() => {
    if (!ticks || !rect.width) {
      return {};
    }

    return Object.keys(ticks).reduce((acc, number) => {
      const absoluteProportion = (Number(number) - min) / Math.abs(min - max);
      return {
        ...acc,
        [number]: {
          position: lerp(0, rect.width, absoluteProportion),
          label: ticks[number],
        },
      };
    }, {});
  }, [ticks, rect.width]);

  // Event handlers
  function handleMouseDown(e) {
    if (e.button !== 0) {
      return;
    }
    const mousePosition = e.pageX;
    setPositions(mousePosition);
    setMoving('mouse');
  }

  function handleMouseMove(e) {
    const mousePosition = e.pageX;
    setPositions(mousePosition);
  }

  function handleTouchStart(e) {
    if (isNotTouchEvent(e)) {
      return;
    }
    const position = e.touches[0].pageX;
    setPositions(position);
    setMoving('touch');
  }

  function handleTouchMove(e) {
    if (isNotTouchEvent(e)) {
      return;
    }
    const position = e.touches[0].pageX;
    setPositions(position);
  }

  function handleEnd() {
    setMoving(null);
  }

  function handleKeyDown(e) {
    // preventDefault here prevents the parent window
    // from scrolling when interacting with the slider

    onKeys(['arrowLeft', 'arrowDown'], () => {
      e.preventDefault();
      setValue(sliderValue - interval);
    })(e);

    onKeys(['arrowRight', 'arrowUp'], () => {
      e.preventDefault();
      setValue(sliderValue + interval);
    })(e);

    onKey('home', () => {
      e.preventDefault();
      setValue(min);
    })(e);

    onKey('end', () => {
      e.preventDefault();
      setValue(max);
    })(e);
  }

  // Sets slider value based on an x position
  function setPositions(xPosition) {
    const clampedPixelOffset = clamp(xPosition - rect.left, 0, rect.width);
    const percentOffset = clampedPixelOffset / rect.width;
    setValue(lerp(min, max, percentOffset));
  }

  // Sets value to a provided value
  function setValue(newValue) {
    const clampedValue = clamp(newValue, min, max);
    setSliderValue(roundToPlaces(clampedValue, precision));
  }

  // Binding of mouse/touch drag events
  React.useEffect(() => {
    if (moving === 'mouse') {
      environment.addEventListener('mousemove', handleMouseMove);
      environment.addEventListener('mouseup', handleEnd);
    }

    if (moving === 'touch') {
      environment.addEventListener('touchmove', handleTouchMove);
      environment.addEventListener('touchend', handleEnd);
    }

    return () => {
      if (moving === 'mouse') {
        environment.removeEventListener('mousemove', handleMouseMove);
        environment.removeEventListener('mouseup', handleEnd);
      }

      if (moving === 'touch') {
        environment.removeEventListener('touchmove', handleTouchMove);
        environment.removeEventListener('touchend', handleEnd);
      }
    };
  }, [moving]);

  const tickMarkup = Object.keys(tickLocations).map((tick) => {
    const { label, position } = tickLocations[tick];

    return (
      <StyledTick
        key={tick}
        style={{ left: position }}
        $disabled={disabled}
        $included={Number(tick) < sliderValue}
      >
        <StyledTickLabel $x={tick}>{label}</StyledTickLabel>
      </StyledTick>
    );
  });

  const assignRefs = (node) => {
    if (sliderRef) {
      sliderRef.current = node;
    }
    if (resizeRef) {
      resizeRef(node);
    }
  };

  return (
    <StyledSlider
      hasTicks={ticks}
      disabled={disabled}
      data-id="slider-wrapper"
      onTouchStart={disabled ? noop : handleTouchStart}
      onMouseDown={disabled ? noop : handleMouseDown}
      ref={assignRefs}
      {...systemProps}
    >
      <StyledRail $disabled={disabled} />
      {tickMarkup}
      <StyledTrack $disabled={disabled} style={{ width: sliderLocation }} />
      <StyledHandle
        id={id}
        aria-controls={props['aria-controls']}
        aria-labelledby={props['aria-labelledby']}
        aria-valuemin={min}
        aria-valuemax={max}
        aria-valuenow={sliderValue}
        aria-disabled={disabled}
        data-id={props['data-id']}
        onBlur={onBlur}
        onFocus={onFocus}
        onKeyDown={disabled ? noop : handleKeyDown}
        role="slider"
        $disabled={disabled}
        style={{ left: sliderLocation }}
        tabIndex={0}
      >
        <StyledHandleShadow $disabled={disabled} />
      </StyledHandle>
    </StyledSlider>
  );
}

Slider.displayName = 'Slider';

Slider.defaultProps = {
  min: 0,
  max: 100,
  precision: 0,
};

export default Slider;
