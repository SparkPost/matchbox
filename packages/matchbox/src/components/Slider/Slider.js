/* eslint max-lines: ["error", 250] */
import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { getRectFor, lerp, useWindowSize } from '../../helpers/geometry';
import { noop, isNotTouchEvent } from '../../helpers/event';
import { onKey, onKeys } from '../../helpers/keyEvents';
import { roundToPlaces, clamp } from '../../helpers/math';
import styles from './Slider.module.scss';

function Slider(props) {
  const { defaultValue, disabled, id, max, min, onBlur, onFocus, onChange, precision, ticks, value } = props;

  const windowSize = useWindowSize(50);
  const [sliderValue, setSliderValue] = React.useState(value || defaultValue);
  const [sliderLocation, setSliderLocation] = React.useState(0);
  const [rect, setRect] = React.useState({});
  const [moving, setMoving] = React.useState();
  const sliderRef = React.useRef();

  React.useEffect(() => {
    setRect(getRectFor(sliderRef.current));
  }, [windowSize, sliderRef.current]);

  // Calculates step increments based on precision
  const interval = React.useMemo(() => {
    let interval = 1;
    if (precision > 0) {
      const zeros = '0'.repeat(precision - 1);
      interval = parseFloat(`0.${zeros}1`);
    }
    return interval;
  }, [precision]);

  // Sets internal value when value is controlled externally
  React.useEffect(() => {
    if (!isNaN(parseFloat(value)) && isFinite(value)) {
      setValue(value);
    }
  }, [value]);

  // Updates slider location when value changes
  React.useEffect(() => {
    if (rect.width) {
      const absoluteProportion = (sliderValue + Math.abs(min)) / Math.abs(min - max);
      setSliderLocation(lerp(0, rect.width, absoluteProportion));
      if (onChange) {
        onChange(sliderValue);
      }
    }
  }, [sliderValue, rect.width]);

  // Updates tick locations when ticks or window size change
  const tickLocations = React.useMemo(() => {

    if (!ticks || !rect.width) {
      return {};
    }

    return Object.keys(ticks).reduce((acc, number) => {
      const absoluteProportion = (Number(number) + Math.abs(min)) / Math.abs(min - max);
      return {
        ...acc,
        [number]: {
          position: lerp(0, rect.width, absoluteProportion),
          label: ticks[number]
        }
      };
    }, {});

  }, [ticks, rect.width]);

  // Event handlers
  function handleMouseDown(e) {
    if (e.button !== 0) {
      return;
    }
    const mousePosition = e.pageX;
    setPositions(mousePosition);
    setMoving('mouse');
  }

  function handleMouseMove(e) {
    const mousePosition = e.pageX;
    setPositions(mousePosition);
  }

  function handleTouchStart(e) {
    if (isNotTouchEvent(e)) {
      return;
    }
    const position = e.touches[0].pageX;
    setPositions(position);
    setMoving('touch');
  }

  function handleTouchMove(e) {
    if (isNotTouchEvent(e)) {
      return;
    }
    const position = e.touches[0].pageX;
    setPositions(position);
  }

  function handleEnd(e) {
    setMoving(null);
  }

  function handleKeyDown(e) {
    e.stopPropagation();
    e.preventDefault();

    onKeys(['arrowLeft', 'arrowDown'], () => setValue(sliderValue - interval))(e);
    onKeys(['arrowRight', 'arrowUp'], () => setValue(sliderValue + interval))(e);
    onKey('home', () => setValue(min))(e);
    onKey('end', () => setValue(max))(e);
  }

  // Sets slider value based on an x position
  function setPositions(xPosition) {
    const clampedPixelOffset = clamp(xPosition - rect.left, 0, rect.width);
    const percentOffset = clampedPixelOffset / rect.width;
    setValue(lerp(min, max, percentOffset));
  }

  // Sets value to a provided value
  function setValue(newValue) {
    const clampedValue = clamp(newValue, min, max);
    setSliderValue(roundToPlaces(clampedValue, precision));
  }

  // Binding of mouse/touch drag events
  React.useEffect(() => {
    if (moving === 'mouse') {
      window.addEventListener('mousemove', handleMouseMove);
      window.addEventListener('mouseup', handleEnd);
    }

    if (moving === 'touch') {
      window.addEventListener('touchmove', handleTouchMove);
      window.addEventListener('touchend', handleEnd);
    }

    return (() => {
      if (moving === 'mouse') {
        window.removeEventListener('mousemove', handleMouseMove);
        window.removeEventListener('mouseup', handleEnd);
      }

      if (moving === 'touch') {
        window.removeEventListener('touchmove', handleTouchMove);
        window.removeEventListener('touchend', handleEnd);
      }
    });
  }, [moving]);

  const sliderClasses = classnames(
    styles.Slider,
    disabled && styles.Disabled
  );

  const tickMarkup = Object.keys(tickLocations).map((tick) => {
    const { label, position } = tickLocations[tick];
    return (
      <div
        key={tick}
        style={{ left: position }}
        className={classnames(styles.Tick, tick < sliderValue && styles.Included)}>
        <div className={styles.TickLabel}>{label}</div>
      </div>
    );
  });

  return (
    <div
      className={sliderClasses}
      onTouchStart={disabled ? noop : handleTouchStart}
      onMouseDown={disabled ? noop : handleMouseDown}
      ref={sliderRef}
    >
      <div className={styles.Rail} />
      {tickMarkup}
      <div
        className={styles.Track}
        style={{ width: sliderLocation }}
      />
      <div
        id={id}
        aria-valuemin={min}
        aria-valuemax={max}
        aria-valuenow={sliderValue}
        aria-disabled={disabled}
        className={styles.Handle}
        onBlur={onBlur}
        onFocus={onFocus}
        onKeyDown={disabled ? noop : handleKeyDown}
        role='slider'
        style={{ left: sliderLocation }}
        tabIndex='0'
      />
    </div>
  );
}

Slider.defaultProps = {
  defaultValue: 0,
  min: 0,
  max: 100,
  precision: 0
};

Slider.propTypes = {
  /**
   * The slider's initial value on first render
   */
  defaultValue: PropTypes.number,
  /**
   * Disables focus, key down, mouse and touch events
   */
  disabled: PropTypes.bool,
  /**
   * The slider's lower bounds
   */
  min: PropTypes.number,
  /**
   * The slider's upper bounds
   */
  max: PropTypes.number,
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  onFocus: PropTypes.func,
  /**
   * The number of decimal places to round values to
   */
  precision: PropTypes.number,

  /**
   * Generates tick marks
   */
  ticks: PropTypes.objectOf(PropTypes.node),
  /**
   * A value to programatically control the slider
   */
  value: PropTypes.number
};

export default Slider;
