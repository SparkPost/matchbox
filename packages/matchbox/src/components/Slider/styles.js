import { tokens } from '@sparkpost/design-tokens';
import { StyledSlider } from './Slider';

export const slider = props => {
  let cursor = 'pointer';

  if (props.disabled) {
    cursor = 'not-allowed';
  }

  return `
    padding: ${tokens.spacing_600} 0;
    position: relative;
    cursor: ${cursor};
  `;
};

export const rail = () => `
  position: absolute;
  left: ${tokens.spacing_0};
  right: ${tokens.spacing_0};

  height: ${tokens.spacing_100};
  background: ${tokens.color_gray_300};
  border-radius: ${tokens.borderRadius_200};
  transition: ${tokens.motionDuration_medium} background;
  ${StyledSlider}:hover & {
    background: ${tokens.color_gray_400};
  }
`;

export const track = props => {
  let background = tokens.color_blue_700;

  if (props.disabled) {
    background = tokens.color_gray_600;
  }

  return `
    position: absolute;
    left: ${tokens.spacing_0};
    height: ${tokens.spacing_100};
    background: ${background};
    border-radius: ${tokens.borderRadius_200};
  `;
};

export const tick = props => {
  let background = tokens.color_gray_300;

  if (props.disabled && props.included) {
    background = tokens.color_gray_600;
  }

  if (!props.disabled && props.included) {
    background = tokens.color_blue_700;
  }

  return `
    position: absolute;
    height: ${tokens.spacing_200};
    width: ${tokens.spacing_200};
    border-radius: ${tokens.borderRadius_circle};
    transform: translate(-2px, -2px);

    background: ${background};
    transition: ${tokens.motionDuration_medium} background;

    user-select: none;
  `;
};

const sharedStyles = `
  position: absolute;
  width: ${tokens.spacing_400};
  height: ${tokens.spacing_400};
  border-radius: ${tokens.borderRadius_circle};
  outline: none;
`;

export const handle = props => {
  let border = `3px solid ${tokens.color_blue_700}`;

  if (props.disabled) {
    border = `3px solid ${tokens.color_gray_600}`;
  }

  return `
    ${sharedStyles}
    transform: translate(-7px, -5px);
    border: ${border};
    background: ${tokens.color_white};
    &:focus > div, &:active > div, &:hover > div {
      opacity: 0.3;
      transition: ${tokens.motionDuration_fast} opacity;
    }
  `;
};

export const handleShadow = props => {
  let boxShadow = `none`;

  if (!props.disabled) {
    boxShadow = `0 0 0 5px ${tokens.color_blue_700}`;
  }

  return `
    ${sharedStyles}
    transform: translate(-3px, -3px);
    border: 3px solid transparent;
    background: transparent;
    box-shadow: ${boxShadow};
    transition: ${tokens.motionDuration_fast} opacity;
    opacity: 0;
  `;
};
