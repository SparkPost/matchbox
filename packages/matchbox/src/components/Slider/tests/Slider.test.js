import React from 'react';
import Slider from '../Slider';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import * as geometry from '../../../helpers/geometry';

const windowEvents = {};
global.addEventListener = jest.fn((event, cb) => {
  windowEvents[event] = cb;
});

global.removeEventListener = jest.fn((event, cb) => {
  delete windowEvents[event];
});

describe('Slider component', () => {
  let onChange;

  beforeEach(() => {
    onChange = jest.fn();
    geometry.getRectFor = jest.fn(() => ({
      left: 0, right: 0, width: 200
    }));
  });
  const subject = (props = {}) => mount(<Slider {...props}/>);

  it('should render default state correctly', () => {
    expect(subject()).toMatchSnapshot();
  });

  it('should render a default value and id', () => {
    const slider = subject({ defaultValue: 50, id: 'test-id' });
    expect(slider.find('.Track')).toHaveAttributeValue('style', { width: '100px' });
    expect(slider.find('.Handle')).toHaveAttributeValue('style', { left: '100px' });
    expect(slider.find('.Handle')).toHaveAttributeValue('id', 'test-id');
  });

  it('should handle a provided value', () => {
    const slider = subject({ value: 50, onChange });
    expect(slider.find('.Track')).toHaveAttributeValue('style', { width: '100px' });
    expect(slider.find('.Handle')).toHaveAttributeValue('style', { left: '100px' });
    expect(slider.find('.Handle')).toHaveAttributeValue('aria-valuenow', '50');
    expect(onChange).toHaveBeenCalledWith(50);
  });

  it('should handle a mouse down', () => {
    const slider = subject({ value: 50, onChange });
    slider.find('.Slider').simulate('mouseDown', {
      pageX: 25,
      button: 0
    });
    expect(slider.find('.Track')).toHaveAttributeValue('style', { width: '26px' });
    expect(slider.find('.Handle')).toHaveAttributeValue('style', { left: '26px' });
    expect(onChange).toHaveBeenCalledWith(13);
  });

  it('should handle a mouse drag chain of events', () => {
    const slider = subject({ value: 50, onChange });
    slider.find('.Slider').simulate('mouseDown', { pageX: 0, button: 0 });

    act(() => {
      windowEvents.mousemove({ pageX: 150 });
    });

    slider.update();
    expect(slider.find('.Track')).toHaveAttributeValue('style', { width: '150px' });
    expect(slider.find('.Handle')).toHaveAttributeValue('style', { left: '150px' });
    expect(onChange).toHaveBeenCalledWith(75);

    act(() => {
      windowEvents.mouseup();
    });

    slider.update();
    expect(windowEvents.mousemove).toBe(undefined);
    expect(windowEvents.mouseup).toBe(undefined);
  });

  it('should handle a touch start', () => {
    const slider = subject({ value: 50, onChange });
    slider.find('.Slider').simulate('touchStart', {
      touches: [{ pageX: 25 }]
    });
    expect(slider.find('.Track')).toHaveAttributeValue('style', { width: '26px' });
    expect(slider.find('.Handle')).toHaveAttributeValue('style', { left: '26px' });
    expect(onChange).toHaveBeenCalledWith(13);
  });

  describe('key events', () => {

    it('should handle a increment', () => {
      const slider = subject({ value: 50, onChange });
      slider.find('.Handle').simulate('keyDown', {
        key: 'ArrowUp',
        shiftKey: false
      });
      expect(slider.find('.Track')).toHaveAttributeValue('style', { width: '102px' });
      expect(slider.find('.Handle')).toHaveAttributeValue('style', { left: '102px' });
      expect(onChange).toHaveBeenCalledWith(51);

      slider.find('.Handle').simulate('keyDown', {
        key: 'ArrowRight',
        shiftKey: false
      });
      expect(slider.find('.Track')).toHaveAttributeValue('style', { width: '104px' });
      expect(slider.find('.Handle')).toHaveAttributeValue('style', { left: '104px' });
      expect(onChange).toHaveBeenCalledWith(52);
    });

    it('should handle a decrement with a precision value to the tenth', () => {
      const slider = subject({ value: 50, onChange, precision: 1 });
      slider.find('.Handle').simulate('keyDown', {
        key: 'ArrowDown',
        shiftKey: false
      });
      expect(slider.find('.Track')).toHaveAttributeValue('style', { width: '99.8px' });
      expect(slider.find('.Handle')).toHaveAttributeValue('style', { left: '99.8px' });
      expect(onChange).toHaveBeenCalledWith(49.9);

      slider.find('.Handle').simulate('keyDown', {
        key: 'ArrowLeft',
        shiftKey: false
      });
      expect(slider.find('.Track')).toHaveAttributeValue('style', { width: '99.6px' });
      expect(slider.find('.Handle')).toHaveAttributeValue('style', { left: '99.6px' });
      expect(onChange).toHaveBeenCalledWith(49.8);
    });

    it('should handle home and end key events', () => {
      const slider = subject({ value: 50, onChange });
      slider.find('.Handle').simulate('keyDown', {
        key: 'Home',
        shiftKey: false
      });
      expect(slider.find('.Track')).toHaveAttributeValue('style', { width: '0px' });
      expect(slider.find('.Handle')).toHaveAttributeValue('style', { left: '0px' });
      expect(onChange).toHaveBeenCalledWith(0);

      slider.find('.Handle').simulate('keyDown', {
        key: 'End',
        shiftKey: false
      });
      expect(slider.find('.Track')).toHaveAttributeValue('style', { width: '200px' });
      expect(slider.find('.Handle')).toHaveAttributeValue('style', { left: '200px' });
      expect(onChange).toHaveBeenCalledWith(100);
    });
  });

  describe('disabled', () => {
    it('should render disabled', () => {
      const slider = subject({ value: 50, disabled: true });
      expect(slider.find('.Slider')).toHaveAttributeValue('class', 'Slider Disabled');
      expect(slider.find('.Handle')).toHaveAttributeValue('aria-disabled', 'true');
    });

    it('should not move when disabled', () => {
      const slider = subject({ value: 50, onChange, disabled: true });
      slider.find('.Handle').simulate('keyDown', { key: 'Home', shiftKey: false });
      slider.find('.Slider').simulate('touchStart', { touches: [{ pageX: 25 }]});
      slider.find('.Slider').simulate('mouseDown', { pageX: 25, button: 0 });
      expect(onChange).toHaveBeenCalledTimes(1);
    });
  });

  describe('ticks', () => {
    it('should render ticks', () => {
      const slider = subject({ value: 50, ticks: { 25: 'test tick', 55: 'not included tick' }});
      expect(slider.find('.Tick')).toMatchSnapshot();
    });
  });
});
