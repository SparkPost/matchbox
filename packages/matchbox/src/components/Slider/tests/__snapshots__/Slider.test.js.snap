// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider component should render default state correctly 1`] = `
<Slider
  defaultValue={0}
  max={100}
  min={0}
  precision={0}
>
  <div
    className="Slider"
    onMouseDown={[Function]}
    onTouchStart={[Function]}
  >
    <div
      className="Rail"
    />
    <div
      className="Track"
      style={
        Object {
          "width": 0,
        }
      }
    />
    <div
      aria-valuemax={100}
      aria-valuemin={0}
      aria-valuenow={0}
      className="Handle"
      onKeyDown={[Function]}
      role="slider"
      style={
        Object {
          "left": 0,
        }
      }
      tabIndex="0"
    />
  </div>
</Slider>
`;

exports[`Slider component ticks should render ticks 1`] = `
Array [
  <div
    className="Tick Included"
    key="25"
    style={
      Object {
        "left": 50,
      }
    }
  >
    <div
      className="TickLabel"
    >
      test tick
    </div>
  </div>,
  <div
    className="Tick"
    key="55"
    style={
      Object {
        "left": 110.00000000000001,
      }
    }
  >
    <div
      className="TickLabel"
    >
      not included tick
    </div>
  </div>,
]
`;
